Integrations/src/main/java/io/deephaven/integrations/python/PythonDeephavenSession.java

```diff
@@ -295,34 +274,49 @@ private Object maybeUnwrap(PyObject o) {
     }
 
     @Override
-    public Set<String> getVariableNames() {
-        return Collections.unmodifiableSet(scope.getKeys().collect(Collectors.toSet()));
+    protected Set<String> getVariableNames(Predicate<String> allowName) {
+        return PyLib.ensureGil(() -> scope.getKeys().filter(allowName).collect(Collectors.toUnmodifiableSet()));
     }
 
     @Override
-    public boolean hasVariableName(String name) {
+    protected boolean hasVariable(String name) {
         return scope.containsKey(name);
     }
 
     @Override
-    public synchronized void setVariable(String name, @Nullable Object newValue) {
-        final PyDictWrapper globals = scope.mainGlobals();
-        if (newValue == null) {
-            try {
-                globals.delItem(name);
-            } catch (KeyError key) {
-                // ignore
-            }
-        } else {
-            if (!(newValue instanceof PyObject)) {
-                newValue = PythonObjectWrapper.wrap(newValue);
+    protected synchronized Object setVariable(String name, @Nullable Object newValue) {
+        Object old = PyLib.ensureGil(() -> {
+            final PyDictWrapper globals = scope.mainGlobals();
+
+            if (newValue == null) {
+                try {
+                    return globals.unwrap().callMethod("pop", name);
+                } catch (KeyError key) {
+                    return null;
+                }
+            } else {
+                Object wrapped;
+                if (newValue instanceof PyObject) {
+                    wrapped = newValue;
+                } else {
+                    wrapped = PythonObjectWrapper.wrap(newValue);
+                }
+                // This isn't thread safe, we're relying on the GIL being kind to us (as we have historically done).
+                // There
+                // is no built-in for "replace a variable and return the old one".
+                Object prev = globals.get(name);
+                globals.setItem(name, wrapped);
+                return prev;
             }
-            globals.setItem(name, newValue);
-        }
+        });
 
         // Observe changes from this "setVariable" (potentially capturing previous or concurrent external changes from
         // other threads)
         observeScopeChanges();
+
+        // This doesn't return the same Java instance of PyObject, so we won't decref it properly, but
```

Comment:
I want to make sure I understand this, let's discuss.

Reply:
Call `ScriptSessionQueryScope.putParam("foo", somePyObjectInstance)`, and then call `same = readParamValue("foo")` to get it back - the PyObject instances are pointing at the "same" python object, but `same == somePyObjectInstance` is false - they are separate java object instances, and contain separate borrowed pointers into py/c.

Reply:
What's the solution? Is @jmao-denver aware?

Reply:
There isn't a great solution in general, except to cleanly close PyObject basically everywhere - java gc is required to mop up all the unclosed PyObject. This is wrong in enough places in DHC that it isn't worth dedicating effort to it here.